force-app/main/default/classes/SOQL.cls

```diff
@@ -287,21 +304,23 @@ public with sharing class SOQL {
      * @return A list of Accounts with matching email domains.
      */
     public static List<Account> findAccountsByLeadEmailDomains(Set<Id> leadIds) {
-        // Query the Leads
-
-        // Create a set of email domains from the Leads
-
-        // Loop through each Lead
-
-            // Get the email from the Lead
-
-            // Check if the email is not null
-
-                // Add the email domain to the set
-
-        
-        // Query the Accounts
-        return null; // Replace null with the variable you used to store the result
+        List<Lead> leadList = [SELECT Id, LastName, Company, Email FROM Lead WHERE Id IN :leadIds];
+        Set<String> emailDomains = new Set<String>();
+        for(Lead leadEmail: leadList) {
+            String email = leadEmail.Email;
+            if(email != null) {
+                String domain = email.substringAfter('@');
+                emailDomains.add(domain);
+            }
+        }
+        String query = 'SELECT Id, Name, Website FROM Account WHERE ';
+        List<String> domainValue = new List<String>();
+        for(String domain: emailDomains) {
+            domainValue.add('Website LIKE \'%' + domain + '%\'');
+        }
+        query += String.join(domainValue, 'OR');
```

Comment:
This is a creative way to do this. You can also use `LIKE ('%domain1', '%domain2')` so you can use a bind variable of your collection here instead of these literals I just used